### ISSUE TITLE
Primary Building Block fails in Blackboard 3400 

### ISSUE THREAD

**Issue by cbeach47** [openequella/openEQUELLA-blackboard-integration#4]
Observed in Bb Release 3400.0.1-rel.90+b74efee and Release 3400.11.0-rel.2+11d8a47.

Try to do any action (save building block, access links, begin selection session), and you're presented with:
```
An error has occurred
java.lang.NoSuchMethodError: blackboard.platform.blti.BasicLTIDomainConfigManager.save(Lblackboard/data/blti/BasicLTIDomainConfig;)V
at com.tle.blackboard.buildingblock.Configuration.createDomainConfig(Configuration.java:495)
at com.tle.blackboard.buildingblock.Configuration.ensureLtiPlacement(Configuration.java:319)
...
```

Testing goal:  Confirm the building block works on all three types (self hosted, managed hosted, SaaS) running 3400.

** Areas to review: ** 
_From https://community.blackboard.com/docs/DOC-1595-preparing-your-building-blocks-for-learn-saas_ 
- [x] APIs - confirm no private APIs are being used
There are 36 classes used that are not in the API, thus are 'private':  https://docs.google.com/spreadsheets/d/1rFYtLlW9ZEGICkASkMz57II_tBHmRYxGK4l7nvUCzko/
  * 2/25 (CB) As per the comments below, closing this.
- [x] Database - Revise the DB calls.  Consider a more vendor agnostic approach (ideally the v3200, v3300, and v3400 would be making the same DB calls code-wise)
  * The updated code on a separate thread from @abidingotter handles Oracle, SQL Server, and PostgreSQL.  The quickest path appears to be to merge that code into this repo.  We could explore Hibernate (which is the cleaner solution), but it would increase the LoE. 
- [x] Shared Content Folder - Ensure the shared content folder handling is appropriate
  * 2/25 (CB) As per the comments below, closing this.
- [x] Logging Changes - Review how messages are being logged, and ensure the README is up-to-date on reviewing the logs
  * Phase 1 complete - B2 logging level can now be controlled from the Settings panel.
  * Phase 2 would be implementing a plugin-specific logger.
- [ ] Statelessness - ensure state is not relied on.
There is a bit of session state relied on in WrappedUser - and is run in both managed-hosting and SaaS (search logs for `oeqIntegStatelessIssue`).  Delaying for now, might not be a blocker.
  * 2/25 (CB) - still needed to review
- [x] Java 8 - Currently build with Java 8, so should be fine.  Review for due diligence. 
  * 2/25 (CB) As per the comments below, closing this.
- [x] Tomcat 8 / JSP Precompilation - Ensure the JSPs have been precompiled appropriately, and the TLDs are NOT in the WAR.
  * JSPs are being precompiled now, still need to confirm TLDs and hit each JSP live.
  * 2/25 (CB) As per the comments below, closing this.
- [x] Tomcat 8/ bb-context-config.properties - Review the various updates to the configs
  * 2/25 (CB) As per the comments below, closing this.
- [x] Tomcat 8 / web.xml - Revise version and configs
  * 2/25 (CB) As per the comments below, closing this.
- [x] URL Encoding - Confirm how URLs are being handled and see if any of the stored URLs are going to be problematic 
  * 2/25 (CB) As per the comments below, closing this.
- [x] Permissions - Update permissons to follow best practices
  * 2/25 (CB) As per the comments below, closing this.
- [x] UI Original / UI - Review UI impacts, clean up
  * 2/25 (CB) As per the comments below, closing this.

**Comment by abidingotter** [comment]
* APIs - confirm no private APIs are being used. There are 36 classes used that are not in the API, thus are 'private'

I think the best thing to do is to leave the use of private APIs as is.  Blackboard will be taking on support of the B2 and can consult with us on any breaking changes.  I think it would be a greater risk (and level of effort) to change the code than to leave it as-is.

**Comment by abidingotter** [comment]
As discussed, for the database we will continue using the custom SQL for specific databases.  If we intended to support the B2 past the end of the year then it would have been highly preferable to use schema.xml, however the B2 is being phased out and the fewer modifications the better.

**Comment by abidingotter** [comment]
* Shared Content Folder - Ensure the shared content folder handling is appropriate

We are already using `PlugInUtil.getConfigDirectory(BbUtil.VENDOR, BbUtil.HANDLE);` instead of the path of the JAR file.  There is a mention of checking to see if any logging you are doing goes to the "log folder" as well, but we are using the dedicated logging service and not writing to a file, so that isn't an issue. We are ok here.

* Logging Changes - Review how messages are being logged, and ensure the README is up-to-date on reviewing the logs

Interesting that there specifically refer to logging to a file.  I guess the logging service we use is not expected to be used by B2s?  LogServiceFactory is in their public API, so not sure.  In any case, given we are using the public API, we are also ok here.

* Statelessness - ensure state is not relied on.

Trying to understand how/why the user is using the session and can't make sense of it, even though I originally wrote it... 🤦‍♂️ Will dig deeper.

* Java 8

All ok there.  Can easily compile to Java 8 compatibility regardless of which Java version we actually compile it with.

* Tomcat 8 / JSP Precompilation - Ensure the JSPs have been precompiled appropriately, and the TLDs are NOT in the WAR.

Thanks to Chris, JSPs are now pre-compiled, however I noted that the TLDs are being included in the building block.  I guess the way forward is to remove them and test that the B2 settings page, and the edit options of added oEQ content, are still usable.

* Tomcat 8/ bb-context-config.properties 

Nothing to do here unless removal of the TLDs proves problematic.

* Tomcat 8 / web.xml - Revise version and configs

Currently on 2.3, not 3.0 as specified.  Have updated as recommended and will re-test.

* URL Encoding

We don't expect to ever include '{', '|', or '}' in our URLs.  We do send JSON data which includes braces, but that is sent in the post data, not in the query string.

* Permissions

We do not use any of the blocked or "warned" permissions.  If I remember correctly, the list we have is a minimum set to get the B2 to run.  Potentially we could review it, but it seems unnecessary when all the currently requested permissions are ok by Blackboard.

* UI Original / UI

We don't use any of the functions mentioned.  Particularly, we are using bbNG and not bbUI in our JSPs.

**Comment by abidingotter** [comment]
In relation to all the Tomcat points, I have completely removed TLDs from the B2 and added the bb-context-config.properties file (the code in our custom tag was actually empty...).  Also updated the web.xml version to 3 as recommended in the BB article.

Didn't get time to visit the session use.  Will try to look into that tomorrow.

**Comment by cbeach47** [comment]
This fix has now been pushed to `master` via PR #9  . 

`statelessness` is the only possible outstanding issue, but we haven't seen it be problematic yet.
